### ISSUE TITLE
Remove TraversableOnce, CanBuildFrom to support Scala 2.13

### ISSUE THREAD

**Issue by jdegoes** [scalaz/scalaz-zio#123]
With the collections redesign, some things are deprecated or removed. We should find a way to support 2.13 and older versions.

**Comment by sideeffffect** [comment]
[`IterableOnce`](https://scala-lang.org/files/archive/api/2.13.0-M4/scala/collection/IterableOnce.html) is only in 2.13
So `Iterable` as it is present in both [2.12.x](https://scala-lang.org/files/archive/api/2.12.x/scala/collection/Iterable.html) and [2.13.M4+](https://scala-lang.org/files/archive/api/2.13.0-M4/scala/collection/Iterable.html)?

**Comment by AL333Z** [comment]
Iterable may be a good replacement, but we're gonna need a new way to `BuildFrom` as well, since `CanBuildFrom` won't be an option availlable anymore.

Maybe something à la [Data.Buildable](http://hackage.haskell.org/package/buildable-0.1.0.3/docs/Data-Buildable.html)?

**Comment by ktonga** [comment]
We don't need to build it we just need to fold it.

As sad as it sounds I think the best option is to implement a `fold` function in terms of `Iterator`, it has been around since ever so it's the only common interface shared among all collections across all scala versions.
So users will have to call `toIterator` on whatever collection type they have in hand, which I guess is a zero-cost operation.

**Comment by AL333Z** [comment]
Hi @ktonga ,

I completely get your point about the need of a `fold` function in terms of `Iterator`.
But think I'm missing something then.

E.g. how we can implement `traverse` without notion of `build`? Specifically, how we can build an F[B]?

```scala 
final def traverse[E, A, B, F[_]: Iterator](
  in: F[A]
)(fn: A => IO[E, B]): IO[E, F[B]]
```

In the current implementation this is solved by `CanBuildFrom`.

```scala
  final def traverse[E, A, B, M[X] <: TraversableOnce[X]](
    in: M[A]
  )(fn: A => IO[E, B])(implicit cbf: CanBuildFrom[M[A], B, M[B]]): IO[E, M[B]]
```

**Comment by ktonga** [comment]
@AL333Z please disdain my first comment, I amended my mistake in my last comment. We do need to build collections too.

**Comment by sideeffffect** [comment]
There's `BuildFrom`, which is available from both [2.12](http://static.javadoc.io/org.scala-lang.modules/scala-collection-compat_2.12/0.1.1/scala/collection/compat/BuildFrom.html) and [2.13](https://scala-lang.org/files/archive/api/2.13.0-M4/scala/collection/BuildFrom.html)

**Comment by ktonga** [comment]
We are already cross compiling for 2.11 too

**Comment by sideeffffect** [comment]
I'm afraid `BuildFrom` is not available for 2.11, only 2.12+

**Comment by ktonga** [comment]
But I don't think there's any diff in collections between 2.11 and 2.12 tho

**Comment by sideeffffect** [comment]
the `BuildFrom` comes from the 2.12 -> 2.13 compatibility library I mentioned above

**Comment by ktonga** [comment]
Ok. It seems it's not my day for making guesses

**Comment by AL333Z** [comment]
I checked as well, and I got to the conclusion that there's nothing we can use from scala std lib (2.11 -> 2.13) which we can leverage for this duty. 

So I was thinking about building our own, even if this decision is not gonna be easy..
E.g. where to put this? scalaz-zio? or even, scalaz (and then, how can we import that in scalaz-zio, since the dependency is reversed)?

**Comment by ktonga** [comment]
I'm afraid we will need to setup the project to provide per scala version implementations

**Comment by ktonga** [comment]
core has to remain zero-deps, this needs to be solved for people using zio from vanilla scala. This is already solved by the interop module for scalaz users

**Comment by ktonga** [comment]
It seems API for 2.13 collections is not yet definitive, so I would propose to start working on this in a separate long-term branch and don't merge any of this to master until 2.13 is released.
I can create the branch if you agree @jdegoes. WDYT?

**Comment by jdegoes** [comment]
@ktonga Yes, that seems good, although no need to create the branch now unless someone is willing to work on this ticket right away.

Also I feel very strongly that ZIO `core` must have zero dependencies.

In the worst case we can just hard-code to `List`.

**Comment by ktonga** [comment]
> unless someone is willing to work on this ticket right away

I think @AL333Z is. Although there is no need for a new branch if we go the `List` way :slightly_smiling_face: 

> Also I feel very strongly that ZIO core must have zero dependencies.

I already expressed my agreement on that too

> In the worst case we can just hard-code to List.

I'm down for that!

**Comment by AL333Z** [comment]
I think we can just stick to `List`, avoiding the dep from scala-collection-compat.
For users which plan to use scalaz-zio without scalaz this may be enough.

`scalaz` may then bring in a more general solution, which may be addressed separately (a proper replacement for CanBuildFrom?).

**Comment by ktonga** [comment]
Closed by #147
